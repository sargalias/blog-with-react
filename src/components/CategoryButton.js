import React from 'react';

const CategoryButton = ({label, isActive=false, hoverColor}) => (
  <li
    className={`nav-item ${isActive && 'active'}`}
    style={{color: hoverColor}}
  >
    <a
      className="nav-link"
      href="/"
      onClick={(e) => e.preventDefault()}
    >
      {label} {isActive && <span className="sr-only">(current)</span>}
    </a>
  </li>
);

export default CategoryButton;
