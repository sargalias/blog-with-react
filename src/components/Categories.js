import React from 'react';
import CategoryButton from './CategoryButton';

const Categories = ({ categories=[], activeCategory=null }) => (
  <nav className="navbar navbar-expand-lg navbar-dark bg-dark" id="categories">
    <div className="container">
      <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarCategories"
              aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
        <span>All Categories</span>
        <span id="category-toggle-symbol">^</span>
      </button>

      <div className="collapse navbar-collapse" id="navbarCategories">
        <ul className="navbar-nav ml-auto mr-auto">
          <li className="navbar-brand">Categories</li>

          { categories && categories.map(categoryLabel => (
            <CategoryButton
              label={categoryLabel}
              key={categoryLabel}
              isActive={categoryLabel === activeCategory}
              hoverColor='green'
            />)
          )}
        </ul>
      </div>
    </div>
  </nav>
);

export default Categories;
