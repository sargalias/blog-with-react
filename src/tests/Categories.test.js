import React from 'react';
import { shallow } from 'enzyme';
import Categories from '../components/Categories';
import CategoryButton from '../components/CategoryButton';

test('Categories renders correctly with sample props', () => {
  const wrapper = shallow(<Categories categories={['Food', 'Home', 'Travel']} activeCategory={'Home'} />);
  expect(wrapper).toMatchSnapshot();
});

test('Categories renders with correct number of CategoryButton children', () => {
  const categories = ['t1', 't2', 't3'];
  const wrapper = shallow(<Categories categories={categories} />);
  expect(wrapper.find(CategoryButton).length).toBe(categories.length);
});

test('Categories renders with correct CategoryButton names', () => {
  const categories = ['t1', 't2', 't3'];
  const wrapper = shallow(<Categories categories={categories} />);
  const categoryButtons = wrapper.find(CategoryButton);
  categoryButtons.forEach((categoryButton, i) => {
    expect(categoryButton.prop('label')).toBe(categories[i]);
  });
});

test('Categories with activeCategory=false creates children CategoryButton with isActive=false', () => {
  const categories = ['t1', 't2', 't3'];
  const wrapper = shallow(<Categories categories={categories} />);
  const categoryButtons = wrapper.find(CategoryButton);
  categoryButtons.forEach(categoryButton => {
    let { isActive } = categoryButton.props();
    expect(isActive).toBe(false);
  });
});

test('Categories with activeCategory creates children CategoryButton with correct isActive props', () => {
  const activeCategory = 't3';
  const categories = ['t1', 't2', 't3'];
  const wrapper = shallow(<Categories categories={categories} activeCategory={activeCategory} />);
  const categoryButtons = wrapper.find(CategoryButton);
  categoryButtons.forEach(categoryButton => {
    let { label, isActive } = categoryButton.props();
    if (label === activeCategory)
      expect(isActive).toBe(true);
    else
      expect(isActive).toBe(false);
  });
});
