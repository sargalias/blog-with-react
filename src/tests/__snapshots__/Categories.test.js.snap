// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories renders correctly with sample props 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
  id="categories"
>
  <div
    className="container"
  >
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarCategories"
      data-toggle="collapse"
      type="button"
    >
      <span>
        All Categories
      </span>
      <span
        id="category-toggle-symbol"
      >
        ^
      </span>
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarCategories"
    >
      <ul
        className="navbar-nav ml-auto mr-auto"
      >
        <li
          className="navbar-brand"
        >
          Categories
        </li>
        <CategoryButton
          hoverColor="green"
          isActive={false}
          key="Food"
          label="Food"
        />
        <CategoryButton
          hoverColor="green"
          isActive={true}
          key="Home"
          label="Home"
        />
        <CategoryButton
          hoverColor="green"
          isActive={false}
          key="Travel"
          label="Travel"
        />
      </ul>
    </div>
  </div>
</nav>
`;
